### PR TITLE
[move] add standard math library

### PR DESCRIPTION
**Pull request by 0xhawk**
### Description
Adds a math standard library. There are currently four functions; max, min, average, and pow. These will be helpful, especially for DeFi builders on Aptos Network to calculate something.

### Test Plan
In the file `math.move`.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3172)

### CODE REVIEW

File: aptos-move/framework/aptos-stdlib/sources/math.move [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    /// Return the smallest of two numbers.
+    public fun min(a: u64, b: u64): u64 {
+        if (a < b) a else b
+    }
+
+    /// Return the average of two.
+    public fun average(a: u64, b: u64): u64 {
+        (a & b) + (a ^ b) / 2
```
Comment (movekevin) [inline aptos-move/framework/aptos-stdlib/sources/math.move:16]: I wonder if we really need the gas optimization here vs just doing (a + b) / 2

Comment (movekevin) [reply aptos-move/framework/aptos-stdlib/sources/math.move:16]: Especially as a ^ b can overflow

File: aptos-move/framework/aptos-stdlib/sources/math.move [hunk 2]
```diff
@@ -0,0 +1,75 @@
+/// Standard math utilities missing in the Move Language.
+module aptos_std::math {
+
+    /// Return the largest of two numbers.
+    public fun max(a: u64, b: u64): u64 {
```
Comment (wrwg) [inline aptos-move/framework/aptos-stdlib/sources/math.move:5]: What about `u128`? And even if you don't introduce now, perhaps call these here `max_u64` et. al?

Comment (wrwg) [reply aptos-move/framework/aptos-stdlib/sources/math.move:5]: Alternatively, we can also put this into the module name, as in `math64::max` et. al

File: aptos-move/framework/aptos-stdlib/sources/math.move (hunk 1 above)
Comment (0xhawk) [reply aptos-move/framework/aptos-stdlib/sources/math.move]: Got your point

File: aptos-move/framework/aptos-stdlib/sources/math.move (hunk 2 above)
Comment (0xhawk) [reply aptos-move/framework/aptos-stdlib/sources/math.move]: I love the way to split the file into the two

**Review by movekevin — APPROVED** [review]

File: aptos-move/framework/aptos-stdlib/sources/math64.move [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            n
+        } else {
+            let p = pow(n, e / 2);
+            p = p * p;
+            if (e % 2 == 1) {
+                p = p * n;
+                p
+            } else {
+                p
+            }
```
Comment (lightmark) [inline aptos-move/framework/aptos-stdlib/sources/math64.move:33]:
```suggestion
            if (e % 2 == 1) {
                p * p * n
            } else {
                p * p
            }
```
I wonder which saves more gas?

File: aptos-move/framework/aptos-stdlib/sources/math.move (hunk 1 above)
Comment (lightmark) [reply aptos-move/framework/aptos-stdlib/sources/math.move]: I think `a + b` has the overflow risk other than the original solution.
let's do `(a & b) + (a ^ b) >> 1` using bit ops for the whole expression.

**Review by lightmark — CHANGES_REQUESTED** [review]
Synced with @movekevin please fix the average function. Sorry for the inconvenience.

**Review by lightmark — APPROVED** [review]

**Comment by lightmark** [comment]
gonna land it for now. I'll fix it later.

**Pull request merged**